src/onehot.jl

```diff
@@ -95,72 +102,105 @@ Base.argmax(x::OneHotLike; dims = Colon()) =
+ ⋅
  1
- 0
+ ⋅
 
-julia> Flux.onehot(:c, [:a, :b, :c])
-3-element Flux.OneHotVector{3,UInt32}:
- 0
- 0
- 1
+julia> αβγ = (Flux.onehot(0, 0:2), β, Flux.onehot(:z, [:a, :b, :c], :c))  # uses default
+(Bool[1, 0, 0], Bool[0, 1, 0], Bool[0, 0, 1])
+
+julia> hcat(αβγ...)  # preserves sparsity
+3×3 OneHotMatrix(::Vector{UInt32}) with eltype Bool:
+ 1  ⋅  ⋅
+ ⋅  1  ⋅
+ ⋅  ⋅  1
 ```
 """
-function onehot(l, labels)
-  i = something(findfirst(isequal(l), labels), 0)
-  i > 0 || error("Value $l is not in labels")
+function onehot(x, labels)
+  i = something(findfirst(isequal(x), labels), 0)
+  i > 0 || error("Value $x is not in labels")
   OneHotVector{UInt32, length(labels)}(i)
 end
 
-function onehot(l, labels, unk)
-  i = something(findfirst(isequal(l), labels), 0)
-  i > 0 || return onehot(unk, labels)
+function onehot(x, labels, default)
+  i = something(findfirst(isequal(x), labels), 0)
+  i > 0 || return onehot(default, labels)
   OneHotVector{UInt32, length(labels)}(i)
 end
 
 """
-    onehotbatch(ls, labels[, unk...])
+    onehotbatch(xs, labels, [default])
+
+Returns a `OneHotMatrix` where `k`th column of the matrix is [`onehot(ls[k], labels)`](@ref onehot),
+for vector `xs`. This is a sparse matrix, which stores just `Vector{UInt32}` containing the indices.
+
+If one of the inputs in `xs` is not found in `labels`, that column is `onehot(default, labels)`
+if `default` is given, else an error.
 
-Return a `OneHotMatrix` where `k`th column of the matrix is `onehot(ls[k], labels)`.
+If `xs` is a matrix, then the result is an `AbstractArray{Bool, 3}` which is one-hot along
+the first dimension, i.e. `result[:, k...] == onehot(xs[k...], labels)`.
 
-If one of the input labels `ls` is not found in `labels` and `unk` is given,
-return [`onehot(unk, labels)`](@ref) ; otherwise the function will raise an error.
+Matrix multiplication `M * onehotbatch(...)` is performed efficiently, by simply getting
+one element from every row of `M`. Some concatenation and reshape operations preserve onehot-ness.
+`OneHotArray`s can be moved to the GPU, to get for instance `OneHotMatrix(::CuArray{UInt32, 1})`.
 
 # Examples
 ```jldoctest
-julia> Flux.onehotbatch([:b, :a, :b], [:a, :b, :c])
-3×3 Flux.OneHotArray{3,2,Vector{UInt32}}:
- 0  1  0
- 1  0  1
- 0  0  0
+julia> oh = Flux.onehotbatch(collect("abracadabra"), 'a':'e', 'e')
+5×11 OneHotMatrix(::Vector{UInt32}) with eltype Bool:
+ 1  ⋅  ⋅  1  ⋅  1  ⋅  1  ⋅  ⋅  1
+ ⋅  1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  1  ⋅  ⋅
+ ⋅  ⋅  ⋅  ⋅  1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅
+ ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  1  ⋅  ⋅  ⋅  ⋅
+ ⋅  ⋅  1  ⋅  ⋅  ⋅  ⋅  ⋅  ⋅  1  ⋅
+
+julia> reshape(1:15, 3, 5) * oh
+3×11 Matrix{Int64}:
+ 1  4  13  1  7  1  10  1  4  13  1
+ 2  5  14  2  8  2  11  2  5  14  2
+ 3  6  15  3  9  3  12  3  6  15  3
 ```
 """
-onehotbatch(ls, labels, unk...) = batch([onehot(l, labels, unk...) for l in ls])
+onehotbatch(ls, labels, default...) = batch([onehot(l, labels, default...) for l in ls])
 
 """
-    onecold(y[, labels = 1:length(y)])
+    onecold(y, [labels])
```

Comment:
```suggestion
    onecold(y, [labels = 1:size(y, 1)])
```

Reply:
My vote is to say this once, just below. (Partly because I'm never super-sure whether an `=` in the docstring signature means a keyword or a default.)

Reply:
Are we consistent about sequestering kwargs after a `;`? If so, I think there shouldn't be too much ambiguity.

Reply:
Yea I don't mean actual ambiguity in the computer sense. I just mean in the plain-language sense, of quickly giving the right impression, and deciding which paragraph to put how much detail in to be minimally confusing.

Reply:
Personally, I think the docstring headers should be very specific and not require reading the paragraph. I don't mind saying it once, though my vote is for in the header.

Reply:
Maybe it should still not have `[ ]`, since that's another way of saying that there exists a default. (Maybe that's part of what strikes me as keyword-ish.)
```suggestion
    onecold(y, labels = 1:size(y,1))
```